src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/logger.h"
+#include "random/generators.h"
+#include "ssx/future-util.h"
+
+#include <seastar/core/loop.hh>
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg, size_t max_size, std::chrono::milliseconds keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+client_ptr kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+                _evicted_items.push_back(std::move(item));
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        // If the passwords don't match, update the password on the client, so
+        // that it can reconnect.
+        if (it_hash->client->config().scram_password.value() != user.pass) {
+            vlog(plog.debug, "Updating password for user {}", k);
+            it_hash->client->config().scram_password.set_value(user.pass);
```

Comment:
https://github.com/redpanda-data/redpanda/issues/6593

Reply:
If the connection is already established, changing the password won't do anything. But if it disconnects for some reason, then the new password will be used, and everything should continue to work.

If the password was wrong to start with, then I guess it will fail to connect. New requests will trigger a new connection, which should then succeed.